ui test error output doesn't play nicely with "goto file" in vscode

it seems like the source directory in the printout for ui tests uses "fake-test-src-base" rn, e.g.
```
---- [ui] tests/ui/traits/new-solver/prefer-param-env-on-ambiguity.rs stdout ----

error: test compilation failed although it shouldn't!
status: exit status: 1
command: "/home/lcnr/rust6/build/x86_64-unknown-linux-gnu/stage1/bin/rustc" "/home/lcnr/rust6/tests/ui/traits/new-solver/prefer-param-env-on-ambiguity.rs" "-Zthreads=1" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Zdeduplicate-diagnostics=no" "-Cstrip=debuginfo" "--remap-path-prefix=/home/lcnr/rust6/tests/ui=fake-test-src-base" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/home/lcnr/rust6/build/x86_64-unknown-linux-gnu/test/ui/traits/new-solver/prefer-param-env-on-ambiguity" "-A" "unused" "-Crpath" "-Cdebuginfo=0" "-Lnative=/home/lcnr/rust6/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/home/lcnr/rust6/build/x86_64-unknown-linux-gnu/test/ui/traits/new-solver/prefer-param-env-on-ambiguity/auxiliary" "-Ztrait-solver=next"
stdout: none
--- stderr -------------------------------
error[E0283]: type annotations needed: cannot satisfy `T: Foo<'a>`
  --> fake-test-src-base/traits/new-solver/prefer-param-env-on-ambiguity.rs:7:34
   |
LL | impl<'a, T: Bar<'a>> Foo<'a> for T {}
   |                                  ^
   |
   = note: cannot satisfy `T: Foo<'a>`

error: aborting due to previous error

For more information about this error, try `rustc --explain E0283`.
------------------------------------------
```

this prevents ctrl+leftclick in vscode to take me directly to the relevant point in the source of the failing test.

it would be nice to somehowo keep the actual path in the output for ui tests and only normalize it when writing into the stderr file?

see https://rust-lang.zulipchat.com/#narrow/stream/131828-t-compiler/topic/ui.20tests.20.22fake-test-src-base.22 for more details

--- a/src/tools/compiletest/src/header.rs
+++ b/src/tools/compiletest/src/header.rs
@@ -6,6 +6,7 @@ use std::io::BufReader;
 use std::path::{Path, PathBuf};
 use std::process::Command;
 
+use build_helper::ci::CiEnv;
 use tracing::*;
 
 use crate::common::{Config, Debugger, FailMode, Mode, PassMode};
@@ -276,8 +277,12 @@ impl TestProps {
     /// `//[foo]`), then the property is ignored unless `cfg` is
     /// `Some("foo")`.
     fn load_from(&mut self, testfile: &Path, cfg: Option<&str>, config: &Config) {
-        // Mode-dependent defaults.
-        self.remap_src_base = config.mode == Mode::Ui && !config.suite.contains("rustdoc");
+        // In CI, we've sometimes encountered non-determinism related to truncating very long paths.
+        // Set a consistent (short) prefix to avoid issues, but only in CI to avoid regressing the
+        // contributor experience.
+        if CiEnv::is_ci() {
+            self.remap_src_base = config.mode == Mode::Ui && !config.suite.contains("rustdoc");
+        }
 
         let mut has_edition = false;
         if !testfile.is_dir() {

